Use realm logo as browser favicon.

YashRE42 opened the pull request:
Fixes #11752.

alya commented:
Yeah, having it be a user setting under Display settings > Advanced makes sense to me. Something like, "Use organization logo in browser tab", defaulted to unchecked?

alya commented:
Squinting at the gif, does this also remove the red dot when there are unread messages? I was actually thinking that the dot seems unnecessary and a little busy since we also have unread counts.

YashRE42 commented:
no, this does not effect the red dot:
<details>
<summary>
gif
</summary>

![](https://user-images.githubusercontent.com/33805964/139522867-18bf53c1-e2a2-4bfe-b9fe-fbd42804e3f5.gif)

</details>

However, this is the second time I've felt a lot of lag while recording this gif... I'm going to do a profile and compare the before and after of 6608fa1 (favicon: Use day mode realm logo as custom favicon).

alya commented:
Oh, I see it now -- thanks!

YashRE42 commented:
Okay, so, the profiles suggest that there's no real difference from these commits:
`load_and_set_favicon` takes 2.7 ms when not including the custom image, and
`load_and_set_favicon` takes 2.6 ms when including the custom image.


Here are the profiles I generated: [profiles-for-favicon.zip](https://github.com/zulip/zulip/files/7445973/profiles-for-favicon.zip)

In both cases the most expensive steps are `new Blob()` and `URL.createObjectURL()`.
Since we perform these steps anyway, I will proceed with adding a display setting for this.

YashRE42 commented:
(also, you can open the gif in a new tab and zoom in to see things better)

timabbott commented:
I merged the first commit (aka the preparatory refactoring), since we want that regardless of what form we end up doing this feature in.

YashRE42 commented:
@laurynmm would you be up to giving the first commit a review? (3fa4dc2d17868ba570221e54ab942d72698f8485)

laurynmm commented on templates/zerver/help/include/sidebar_index.md:
```diff
@@ -52,6 +52,7 @@
 * [Configure default view](/help/configure-default-view)
 * [Enable full width display](/help/enable-full-width-display)
 * [Display the buddy list on narrow screens](/help/move-the-users-list-to-the-left-sidebar)
+* [Organization logo as favicon](/help/organization-logo-as-favicon)
```
Maybe match this to the header of the `/help` article so that it starts with a verb: 'Use organization logo as favicon'. I checked with the inspector and I think it will still fit in the sidebar as one line (see image).

<img width="270" alt="Screen Shot 2021-11-02 at 4 12 46 PM" src="https://user-images.githubusercontent.com/63245456/139875297-b6982351-6382-450f-8f06-26884984292c.png">

laurynmm commented on templates/zerver/help/organization-logo-as-favicon.md:
```diff
@@ -0,0 +1,13 @@
+
+This is feature has not yet been implemented.
+
+### Move the users list
```
This was probably just a leftover header from copy/paste, but I'd change to 'Set browser favicon' or something along those lines.

laurynmm reviewed (commented):
@YashRE42 - This looks good based on the similar user setting that I implemented last week. The only thing missing is the update to `version.py` with the new `API_FEATURE_LEVEL`. I also had two tiny notes in the `/help` docs, which are going to be fleshed out anyway when the feature is implemented.

alya commented:
@YashRE42 Is this PR ready for me to play with, or not quite?

YashRE42 commented:
It is!

I've also made the changes requested above. CI failed due to puppeteer flake, it's rerunning on this branch. (I'll force push to rerun it again if the GitHub UI here doesn't update).

alya commented:
Looks like this is using the day logo, but I was imagining we'd use the **Organization profile picture**. I think that one would be better because (a) it's square, and (b) it would make this feature work for Zulip Cloud Free orgs, which can't upload the other type of logo.

 What do you guys think?

YashRE42 commented:
I have no issues with making that change. That's actually what #12549 was going to do initially, but we changed since we'd described the feature with "logo" and not "icon", that change never really got feedback.

(to be clear: the code refers to the "Organization profile picture" as "realm_icon", the "Organisation day logo" as "logo" and "Organisation night logo" as "night logo".)

YashRE42 committed:
settings: Add user display setting to enable realm icon as favicon.

This commit adds a setting in the user display settings, under the
advanced section which we will use to enable or disable the "realm
icon as favicon" feature.

It also changes zulip.yaml and creates the relevant help doc.

As the documentation mentions, this feature has not been implemented
yet, this is just the display setting plumbing.

YashRE42 commented:
(wip)

YashRE42 commented:
![](https://user-images.githubusercontent.com/33805964/140730932-52a81b83-666e-42c1-a508-9580abe251b2.gif)

Updated this to use the Organization profile picture.

alya commented:
Nice, looking good!

Do you guys think the term "favicon" is generally accessible? Or should we have the setting be something like, "Use organization profile picture in browser tab"?

timabbott commented:
I found the logo a bit hard to parse visually; I wonder, is it possible to tell whether we should use the organizations night-theme logo or day-theme logo in the favicon slot?  Maybe `prefers-color-scheme`?

timabbott commented:
@andersk can you do a review on this?  It looks generally reasonable to me, but the favicons have a lot of subtle issues with them that I think you understand better than I do.

alya commented:
> I found the logo a bit hard to parse visually; I wonder, is it possible to tell whether we should use the organizations night-theme logo or day-theme logo in the favicon slot? Maybe `prefers-color-scheme`?

See my comment above:

> Looks like this is using the day logo, but I was imagining we'd use the Organization profile picture. I think that one would be better because (a) it's square, and (b) it would make this feature work for Zulip Cloud Free orgs, which can't upload the other type of logo.

timabbott commented:
Yeah, I know that; the "logo" is the wrong shape anyway.  But I'd forgotten that we don't offer an option to upload different background variants of the organization profile picture.  Given that context, this is the best we can do UX-wise.

alya commented:
Ah, I see where the confusion was, OK. :)

YashRE42 commented:
rebased with upstream main.

andersk commented on static/js/favicon.js:
```diff
@@ -52,14 +59,49 @@ export function update_favicon(new_message_count, pm_count) {
+        // the template because we generate an immutable blob from the
+        // template which would not include the rendered image
+        if (user_settings.realm_icon_as_favicon && page_params.realm_icon_url) {
+            if (!image_data_URI || should_update_image_data_URI) {
```
`image_data_URI` → `icon_data_url` (`icon` is more specific, we use lowercase variable names, and “URI” is [not a thing anymore](https://url.spec.whatwg.org/#goals)).

Don’t use an explicitly managed `should_update_image_data_URI` flag. Instead, make it automatic: remember the `realm_icon_url` corresponding to the cached `icon_data_url`, and update them both if `page_params.realm_icon_url !== realm_icon_url`.

andersk commented on static/templates/favicon.svg.hbs:
```diff
@@ -7,11 +7,15 @@
         }
     </style>
+    {{#if image_data_URI}}
+    <image href="{{image_data_URI}}" height="16px" width="16px" />
```
SVG typically uses unitless lengths (16, not 16px).

andersk commented on static/js/favicon.js:
```diff
@@ -52,14 +59,49 @@ export function update_favicon(new_message_count, pm_count) {
+            if (!image_data_URI || should_update_image_data_URI) {
+                const reader = new FileReader();
+                // readAsDataURL converts the blob to base64 and calls onload
+                fetch(page_params.realm_icon_url)
+                    .then((response) => response.blob())
+                    .then((blob) => reader.readAsDataURL(blob));
+                reader.addEventListener("load", () => {
+                    // store image_data_URI
+                    image_data_URI = reader.result;
```
Always prefer `async`/`await` over `.then`.

```js
const response = await fetch(page_params.realm_icon_url);
reader.readAsDataURL(await response.blob());
await new Promise((resolve) => {
    reader.addEventListener("load", () => {
        resolve();
    });
});
icon_data_url = reader.result;
```

This will also let you fix the duplicated code below (since you can conditionally `await`).

YashRE42 commented:
I've rebased to main, resolved the merge conflict and addressed the feedback above. (This is so much cleaner now! Thanks Anders!)

timabbott commented:
This needs a rebase for the API changelog stuff; and then I think mostly just a documentation/strings pass from me and then we can merge it.

andersk commented on templates/zerver/base.html:
```diff
@@ -16,8 +16,12 @@
+        {% if user_profile and user_profile.realm_icon_as_favicon %}
+        <link id="favicon" rel="icon" href="{{ page_params.realm_icon_url }}" />
+        {% else %}
+        <link id="favicon" rel="icon" href="/static/images/favicon.png?v=4" />
```
The primary icon is supposed to be the SVG, not the PNG.

YashRE42 commented:
I've rebased this: renumbered migrations and bumped feature level to 110.
I've also addressed Anders' last comment, and changed from "favicon.png?v=4" to "favicon.svg?v=4".

GitHub Code Scanning / CodeQL has been running for almost 30 minutes now, is it supposed to run that long?

timabbott commented:
> GitHub Code Scanning / CodeQL has been running for almost 30 minutes over an hour now, is it supposed to run that long?

No, they've been having operational problems.  Those checks are rarely helpful, so if it's not running properly, I'd ignore it.

timabbott commented:
I've pushed back here with a small tweak to the label for the setting; @alya will do another pass on the /help/ documentation after we merge this.  

Before merging, I'd like to ask for more detail on the removal of `<link rel="alternate icon" href="/static/images/favicon.png?v=4" />`; I'm pretty sure we added that alternate for a reason, and it seems like that could break older browsers that don't support SVG favicons.  Can we just change the alternate icon URL when changing the favicon configuration as well?

YashRE42 commented:
it... doesn't seem consistent with just clearing the href, for example it would sometimes show the zulip icon when the page first loads or when unreads go to 0, but clearing the href and also inserting the alternate icon only if the setting is off seems to work alright (the risk of course being that if someone on an old browser loads the page with the realm logo as icon setting on and then turns it off, they might see a broken favicon)...

I've pushed with that change, do you feel like it's an acceptable compromise?

andersk commented on static/js/favicon.js:
```diff
@@ -23,7 +27,11 @@ function set_favicon() {
 
-export function update_favicon(new_message_count, pm_count) {
+function clear_alternate_icon() {
+    $("#alternate_icon").attr("href", "");
```
An empty string resolves as a relative URL to the current page. That’s not what you want.

andersk commented on static/js/favicon.js:
```diff
@@ -39,7 +47,16 @@ export function update_favicon(new_message_count, pm_count) {
         if (new_message_count === 0 && pm_count === 0) {
-            $("#favicon").attr("href", "/static/images/favicon.svg?v=4");
+            if (user_settings.realm_icon_as_favicon && page_params.realm_icon_url) {
+                // Having both an alternate_icon and the realm_icon set as the
+                // favicon would sometimes only show the alternate_icon, so we
+                // first clear the alternate icon.
+                clear_alternate_icon();
```
Can you debug this more? Under what circumstances does this happen? Why does this not happen with our existing dynamic favicon? Or does it?

timabbott replied:
Yeah, I'd also prefer that we debug/reproduce this.

YashRE42 replied:
> Or does it?

This was key, I think I'm on track to figuring things out. We don't currently notice this behavior because our png and svg icons look the same.

YashRE42 replied:
huh. or maybe not. I can no longer reproduce any problems if I test with `disable cache` in the network tab checked.

timabbott commented:
The merge conflicts are because we merged an API feature level 110, so this'll need a renumber.

YashRE42 commented:
wip

YashRE42 commented:
This PR works with that last commit, we can squash if necessary?

YashRE42 commented:
rebased and switched to feature level 112.

YashRE42 commented:
Updated implementation to use `update_alternate_icon` as well as `attr("sizes"` and `removeAttr("sizes"`... But I've only tested this with chrome 94...

andersk commented on templates/zerver/base.html:
```diff
@@ -16,8 +16,33 @@
+        zulip svg.
+
+        Also, the above bug may be chrome specific, (see:
+        https://crbug.com/1140920)
```
https://crbug.com/1140920 isn’t related to alternate icon selection. Did you mean to paste a different link?

YashRE42 replied:
Correct, I meant to link https://crbug.com/1162276, but the confusing thing is that the bug report describes the opposite of what we're seeing. It says that "with sizes="any" the .png is always chosen" what we're seeing is that "with sizes="any", the icon is chosen based on last declared order" and "without sizes="any", the .png is always chosen".

andersk commented on templates/zerver/base.html:
```diff
@@ -16,8 +16,33 @@
+        https://crbug.com/1140920)
+        -->
+        {% if user_profile and user_profile.realm_icon_as_favicon %}
+        <link id="alternate_icon" rel="alternate icon" href="{{ page_params.realm_icon_url }}" sizes="any"/>
+        <link id="favicon" rel="icon" href="{{ page_params.realm_icon_url }}" sizes="any"/>
```
Why do we need `sizes="any"`, and why only in this case?

YashRE42 replied:
To the best of my understanding, adding `sizes="any"` is what makes the browser (or at least, chrome) consider these to be "equally appropriate icons". I believe this because without this attribute, we see the bug described above.

In addition, I believe we do not need this attribute with the zulip logo because the svg is already considered more appropriate than the png (which may be because they both have the same dimensions?).

andersk commented on static/js/favicon.js:
```diff
@@ -52,13 +70,45 @@ export function update_favicon(new_message_count, pm_count) {
+        if (user_settings.realm_icon_as_favicon && page_params.realm_icon_url) {
+            // generate new icon_data_url if necessary
+            if (!current_realm_icon_url || current_realm_icon_url !== page_params.realm_icon_url) {
+                const response = await fetch(page_params.realm_icon_url);
```
Waiting for a network request here has some failure modes.

* There could be many updates to the count while the image is still loading. This will generate many image requests, and they will resolve in unclear order, so the count displayed at the end might be wrong.
* The image could fail to load, in which case the count will never be updated at all.

I think we want `update_favicon` to remain synchronous. We want a separate task, triggered by page load or settings updates, that fetches the image, stores the data URL, and does an `update_favicon` with the current counts as of when the image *finished* loading. Then the user will always see an up-to-date count even if the image is stale. This perhaps means the `favicon` module needs to remember the current counts, which would also solve some problems you had elsewhere.

YashRE42 replied:
if you're concerned about that, can we go one step further and precompute the data url on the backend with PIL or something and make that available via page params (like the commit message talked about)? If we go down that path we should obviously add a DB entry rather than computing at runtime... 

The main reason why I'm interested in that direction is because it would mean that we don't need to wait for a request at all in the normal (`update_favicon`) path. We would still need to add a (`update_realm_icon_data_url`) to handle the case of a new realm icon being uploaded (which would only be called by `server_events_dispatch`), but I suspect this would be simpler... would you say what I'm describing is a reasonable plan?

YashRE42 commented:
I attempted the approach I suggested in https://github.com/zulip/zulip/pull/20088#discussion_r783661859. I didn't end up needing to implement an additional `update_realm_icon_data_url`. The net changes to `favicon.js` are much smaller now.

CI fails because I have not updated backend tests, but if someone could have a look and validate this approach, I could do that next. (I have the previous approach saved as a backup.)

cc: @andersk.

timabbott commented on static/js/server_events_dispatch.js:
```diff
@@ -286,6 +288,15 @@ export function dispatch_normal_event(event) {
+                                // because favicon.js does not keep track of unreads,
+                                // once they have been rendered.
+                                notifications.redraw_favicon();
+                                // incase we fail to render the svg
```
Can you explain this possibility in more detail?

Also, should `notifications.redraw_favicon` just be responsible for doing `favicon.update_attrs_and_alt_icon` instead?

timabbott commented:
Hmm. I don't like the idea of sending a base-64 encoded realm icon content in `page_params`. I could see it being justified for clients that have this setting enabled, if doing so is required if we want to be able to set the favicon immediately on page load, rather than having it flash the zulip logo. Is that an issue we need to work around here?

I think [Anders comment in this thread](https://github.com/zulip/zulip/pull/20088#discussion_r783661859) remains a cleaner way to do this that sort of flicker issue does not create a requirement that the data be in that payload. I don't think it's hard to write a JS function function to fetch the icon from the server, noops if it's already in progress, will retry on `error`, and has a `success` callback that saves the result and then does `notifications.redraw_favicon` to show the proper count.

YashRE42 committed:
WIP.

favicon: Use realm icon as browser favicon.

Previously, it was difficult to distinguish between multiple Zulip
realms, especially when pinned in the browser, as they all used the
same, default Zulip favicon. This commit allows users to set the
favicon to be the same as realm icon. (We still show unread counts
over the favicon, and the red dot in case of pms, in the same manner
as before).

There are a couple of ways to produce the base64 data url, on the
frontend these include making a request to the url using fetch() or
XMLHttpRequest() and then converting to base64 via canvas.toDataURL()
or by FileReader().readAsDataURL().

However, partially due to the fact that a new request would be a point
of failure that we would need to account for, and partly due to the
fact that we should simply avoid making the request if we can, we
implement the conversion to base64 on the backend, and just add the
data url to page_params. This is essentially precomputing.

This commit also completes the /help doc describing this user setting.

This is currently not merge ready due to the following:
TODO: Add tests for the backend changes.

Fixes: #11752.

YashRE42 committed:
migration: Renumber migration.

This should be squashed back into the commit which creates this
migration.

YashRE42 committed:
wip: wip.

alya commented:
Thanks! A few notes from testing the PR:

1. The menu dropdown needs to be styled like our other dropdowns.

![Screen Shot 2023-02-24 at 9 16 02 AM](https://user-images.githubusercontent.com/2090066/221244999-6cc127e8-14ac-405e-bf7a-2f803d180822.png)

2. When I set the setting to "Never", the emoji in the sidebars (status) still animate. Is this intentional? I don't recall if we had prior discussions about it.

3. The setting does not work correctly for `/me` messages.

That's all I've got. :)

YashRE42 commented:
This was last blocked on a hard-to-reproduce bug where we weren't rerendering the custom favicon correctly on the first load (and some combination of having a non-zero unread). The last commit here was a WIP attempt at implementing the relevant logic in JS. The next step would be to drop the existing commits that attempted this via upload.py and see if we can figure out a reproducer for the bug.
